### ISSUE TITLE
Support for circular searches in get_events

### ISSUE THREAD

**Issue by calum-chamberlain** [niosh-mining/obsplus#107]
* obsplus version: 0.0.2
* Python version: 3.6.6
* Operating System: Ubuntu 16.04

### Description

I have a need to select data using the `latitude`, `longitude` and `maxradius` kwargs - I thought before I went and hacked something together I would check whether there were any preferred ways of adding that kind of thing? Or if it should always stay out of the scope of obsplus?

### What I Am Thinking of Doing

Calculating a column from the latitude and longitude columns in the dataframe relative to the given latitude and longitude and filtering based on that - it does introduce extra processing time, but calculating the distance for all events relative to some given location might also be helpful for other purposes. 

A possible distance calculation function could look something like:
```python
from obspy.geodetics import gps2dist_azimuth, kilometer2degrees

def calc_dist(lat1, lon1, df, degrees=True): # This should also pass a and r if given.
    def _dist_func(_df):
        dist, _, _ = gps2dist_azimuth(lat1=lat1, lon1=lon1, lat2=_df["latitude"], lon2=_df["longitude"])
        if degrees:
            return kilometer2degrees(dist / 1000.)
        return dist / 1000
    return df.apply(_dist_func, axis=1)
```
this returns a new Series that could be be filtered to give indexes using the `obsplus.events.get_events._get_ids` function with `minradius` and `maxradius` essentailly already supported.  To keep it working with multiple parameters though I think the "radius" column would have to be temporarily added to the dataframe?

**Comment by d-chambers** [comment]
Hey @calum-chamberlain,

This is certainly not outside the scope of ObsPlus and would be a very welcome contribution! The goal is that `EventBank.get_events` method eventually be a superset of obspy's `fdsn.Client.get_events`, so this will certainly help us get a few steps closer. 

I was thinking something similar for the implementation. It would look like this:

1. Determine if the `maxradius` keyward was used, if so check if any of the `minlatitude`, `maxlatitude`, `minlongitude`, `maxlongitude` keywords were used and raise a `ValueError` if so because these keywords are not compatible. 

2. Determine the `minlatitude`, `maxlatitude`, `minlongitude`, `maxlongitude` of a square which circumscribes, or at lease safely exceeds, the bounds of the circle defined by `maxradius`. Pass these, and any other query parameters to a separate `read_index` call and return the dataframe.

3. Calculate the radius from `latitude` and `longitude` for each row and add a `radius` column

4. Filter out any rows that exceed `maxradius`

5. Filter out any rows that don't exceed `minradius`, if `minradius` is not `None`

6. Return resulting dataframe and maybe drop the radius column, unless you think it would be useful to keep. An extra column usually wont break anything; it's removing expected columns that can cause havoc. 

According to [FDSN Client docs](https://docs.obspy.org/packages/autogen/obspy.clients.fdsn.client.Client.get_events.html) minradius and maxradius should be in degrees. It wouldn't be bad to also support using great circle distance, however. I am not sure the best way to do that though, maybe a `maxradius_m` parameter?

**Comment by calum-chamberlain** [comment]
Jolly good: I have started working on this on [this branch](https://github.com/calum-chamberlain/obsplus/tree/circular-search) of my fork. It seems to be working at the moment, but I'm going to play around with some more tests.  It also isn't necessarily the cleanest implementation. Darned good idea to add the rectangular filter first, speed things up nicely.

I ended up changing a few more things than I anticipated. A few questions I have are:
1. I added a [`calculate_distance`](https://github.com/calum-chamberlain/obsplus/blob/circular-search/obsplus/events/get_events.py#L106) function, but didn't know where that should be put. At the moment it is in `obsplus.events.get_events`
2. `EventBank.read_index` set the index of the dataframe to be `event_id`.  This clashed with the `obsplus.events.get_events._get_ids` function, so I moved the `.set_index("event_id")` to the end of `_read_index`.  I wonder if it would be better to set the index in `obsplus.events.pd.events_to_df` so that the dataframes are consistently indexed? I'm probably missing something here?
3. I didn't end up adding the radius column to the dataframe, I just used it as a series. I'm kinda keen to keep the internal dataframe representation consistent.

**Comment by d-chambers** [comment]
> I added a calculate_distance function, but didn't know where that should be put. At the moment it is in obsplus.events.get_events

We have taken to a distributed utils approach, meaning there is an utils modules sprinkled about obsplus rather than one utils module with a lot of sub-modules. So the best place for this function is probably `obsplus.events.utils`, unless you think it might be used for inventories as well, in which case it can be elevated to `obsplus.utils`.

> EventBank.read_index set the index of the dataframe to be event_id. This clashed with the obsplus.events.get_events._get_ids function, so I moved the .set_index("event_id") to the end of _read_index. I wonder if it would be better to set the index in obsplus.events.pd.events_to_df so that the dataframes are consistently indexed? I'm probably missing something here?

This was more organic rather than a deliberate. Let me think about it but I know we have a fair bit of code which relies on event_id being a column. In a sense it might be better that way, then the user can set whatever columns as indices as he or she deems appropriate for the task at hand without having to reset the index. `EventBank` does set the dataframe index to the event_id, however, because I thought it would be more efficient if SQLite used event_id as its index. However, I am not sure if `DataFrame.to_sql` is really honoring the index at all; I haven't looked into it.  It does bother me a bit that the dataframes are not consistently indexed but `EventBank`'s index should, ideally, only contain as much information as is necessary to perform all the supported queries so the two dataframes may diverge a bit. Keeping `EventBank`s index a subset of the output of `events_to_df` wouldn't be a bad idea though.  

> I'm kinda keen to keep the internal dataframe representation consistent. 

Yes, good idea.

**Comment by calum-chamberlain** [comment]
Thanks for that: 
- [ ] I will move `calculate_distance` - good point on using for inventories, I think it would be useful for that purpose as well;
- [ ] RE indexing: fair enough, as long as you are okay with me setting the index at the end of the `_get_ids` method then I think the current way of working works...

Extra thought, reading events in `EventBank.get_events` doesn't need to be done in serial AFAIK - thoughts on having an `executor` as either a supported kwarg for some functions, or as a class attribute that if it exists some methods will use it? I played around a bit here and it sped up my event reading nicely, but it probably won't be useful for many people. I also didn't do any timing, but can do if it fits.  I'm cagey about adding parallel things either at all, or prematurely (but having `concurrent.futures` in Python 3 is excellent!).

**Comment by d-chambers** [comment]
Ok, @sboltz any thoughts on the index of the output dataframe of `events_to_df` and the dataframe used by `EventBank`? Should we make them the same and if so which way would be better? 

I think allowing concurrency is a good idea; it may even be able to speed up indexing for both `EventBank` and `WaveBank`. Whatever we do here it would be good if it was compatible with [dask](https://distributed.dask.org/en/latest/), which advertises `concurrent.futures` compatibility so that is probably the way to go. Let me give it a bit of thought though before I send you down the wrong path again. I do think it should be an opt-in feature rather than the default.

**Comment by sboltz** [comment]
I am definitely in favor of standardizing it, as that has tripped me up before. As for which it use, I kind of lean more towards leaving ‘event_id’ as a column, both to minimize changing existing code and to be consistent with the other extractors, some of which can’t always retrieve an ‘event_id’ (specifically ‘arrivals_to_df’).

**Comment by d-chambers** [comment]
closed by #110
